src/Agent.Worker/StepsRunner.cs

```diff
@@ -72,36 +73,41 @@ public sealed class StepsRunner : AgentService, IStepsRunner
                 step.ExecutionContext.Variables.RecalculateExpanded(out expansionWarnings);
                 expansionWarnings?.ForEach(x => step.ExecutionContext.Warning(x));
 
+                // Record the job cancellation state prior evaluating the condition.
+                bool alreadyCanceled = jobContext.CancellationToken.IsCancellationRequested;
+
                 // Evaluate condition.
                 step.ExecutionContext.Debug($"Evaluating condition for step: '{step.DisplayName}'");
                 var expressionManager = HostContext.GetService<IExpressionManager>();
+                DT.INode conditionTree = null;
                 bool? conditionResult = null;
                 try
                 {
-                    conditionResult = expressionManager.Evaluate(step.ExecutionContext, step.Condition);
+                    conditionTree = expressionManager.Parse(step.ExecutionContext, step.Condition);
```

Comment:
I looked into this some more. Moving all condition validation up front before any step is started is a bigger change. It's not huge, but it's a pattern change I want to talk through it before going after that change.

Also I have changed my mind that condition evaluation failure should be treated as critical. Let's chat Monday, I want to get your thoughts.